docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,334 @@
+a script to inspect data as it is streaming by.
+
+.. code-block:: lua
+
+  for chunk in request_handle:bodyChunks() do
+    request_handle:log(0, chunk:byteSize())
+  end
+
+Each chunk the iterator returns is a :ref:`buffer object <config_http_filters_lua_buffer_wrapper>`.
+
+trailers()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  trailers = handle:trailers()
+
+Returns the stream's trailers. May return nil if there are no trailers. The trailers may be
+modified before they are sent to the next filter.
+
+Returns a :ref:`header object <config_http_filters_lua_header_wrapper>`.
+
+log*()
+^^^^^^
+
+.. code-block:: lua
+
+  handle:logTrace(message)
+  handle:logDebug(message)
+  handle:logInfo(message)
+  handle:logWarn(message)
+  handle:logErr(message)
+  handle:logCritical(message)
+
+Logs a message using Envoy's application logging. *message* is a string to log.
+
+httpCall()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  headers, body = handle:httpCall(cluster, headers, body, timeout)
+
+Makes an HTTP call to an upstream host. Envoy will yield the script until the call completes or
+has an error. *cluster* is a string which maps to a configured cluster manager cluster. *headers*
+is a table of key/value pairs to send. Note that the *:method*, *:path*, and *:authority* headers
+must be set. *body* is an optional string of body data to send. *timeout* is an integer that
+specifies the call timeout in milliseconds.
+
+Returns *headers* which is a table of response headers. Returns *body* which is the string response
+body. May be nil if there is no body.
+
+respond()
+^^^^^^^^^^
+
+.. code-block:: lua
+
+  handle:respond(headers, body)
+
+Respond immediately and do not continue further filter iteration. This call is *only valid in
+the request flow*. *headers* is a table of key/value pairs to send. Note that the *:status* header
+must be set. *body* is a string and supplies the optional response body. May be nil.
+
+.. _config_http_filters_lua_header_wrapper:
+
+Header object API
+-----------------
+
+add()
+^^^^^
+
+.. code-block:: lua
+
+  headers:add(key, value)
+
+Adds a header. *key* is a string that supplies the header key. *value* is a string that supplies
+the header value.
+
+get()
+^^^^^
+
+.. code-block:: lua
+
+  headers:get(key)
+
+Gets a header. *key* is a string that suplies the header key. Returns a string that is the header
+value or nil if there is no such header.
+
+__pairs()
+^^^^^^^^^
+
+.. code-block:: lua
+
+  for key, value in pairs(headers) do
+  end
+
+Iterates through every header. *key* is a string that supplies the header key. *value* is a string
+that supplies the header value.
+
+.. attention::
+
+  In the currently implementation, headers cannot be modified during iteration. Additionally, if
+  it is desired to modify headers after iteration, the iteration must be completed. Meaning, do
+  not use `break` or any other mechanism to exit the loop early. This may be relaxed in the future.
```

Comment:
given that we are in scripting land, does it make sense in the initial cut to simply copy out the entire header map to lua, and then copy it back into envoy?
like
```lua
hdrs = headers:clone() // full copy
.. do lua stuff [ipairs/pairs/etc]
headers:reset(hdrs)
```

Reply:
Meh, I think the key/value and direct modification is nicer? I think I can fix any usability issues that people encounter.

Reply:
direct modification is nice. But the text suggests that one cannot modify keys during iteration, and that the iteration has to be completed (not even break).

Reply:
Let me think about this. I could see additionally adding a clone/merge API that operates on tables, but IMO that should be separate.